tools/block-generator/generator/generator_ledger.go

```diff
@@ -0,0 +1,342 @@
+	}, cconfig.GetDefaultLocal())
+	if err != nil {
+		fmt.Printf("error initializing ledger: %v\n.", err)
+		os.Exit(1)
+	}
+	g.ledger = l
+}
+
+func (g *generator) minTxnsForBlock(round uint64) uint64 {
+	// There are no transactions in the 0th round
+	if round == 0 {
+		return 0
+	}
+	return g.config.TxnPerBlock
+}
+
+// startRound updates the generator's txnCounter based on the latest block header's counter.
+// It is assumed that g.round has already been incremented in finishRound()
+func (g *generator) startRound() error {
+	var latestRound basics.Round
+	if g.round > 0 {
+		latestRound = basics.Round(g.round - 1)
+	} else {
+		latestRound = basics.Round(0)
+	}
+
+	latestHeader, err := g.ledger.BlockHdr(latestRound)
+	if err != nil {
+		return fmt.Errorf("could not obtain block header for round %d (latest round %d): %w", g.round, latestRound, err)
+	}
+
+	g.txnCounter = latestHeader.TxnCounter
+	return nil
+}
+
+// finishRound tells the generator it can apply any pending state and updates its round
+func (g *generator) finishRound() {
+	g.timestamp += consensusTimeMilli
+	g.round++
+
+	// Apply pending assets...
+	g.assets = append(g.assets, g.pendingAssets...)
+	g.pendingAssets = nil
+
+	g.latestPaysetWithExpectedID = nil
+	g.latestData = make(map[TxTypeID]uint64)
+
+	for kind, pendingAppSlice := range g.pendingAppSlice {
+		for _, pendingApp := range pendingAppSlice {
+			appID := pendingApp.appID
+			if g.appMap[kind][appID] == nil {
+				g.appSlice[kind] = append(g.appSlice[kind], pendingApp)
+				g.appMap[kind][appID] = pendingApp
+				for sender := range pendingApp.optins {
+					g.accountAppOptins[kind][sender] = append(g.accountAppOptins[kind][sender], appID)
+				}
+			} else { // just union the optins when already exists
+				for sender := range pendingApp.optins {
+					g.appMap[kind][appID].optins[sender] = true
+					g.accountAppOptins[kind][sender] = append(g.accountAppOptins[kind][sender], appID)
+				}
+			}
+		}
+	}
+	g.resetPendingApps()
+}
+
+// ---- ledger block evaluator ----
+
+func (g *generator) startEvaluator(hdr bookkeeping.BlockHeader, paysetHint int) (*eval.BlockEvaluator, error) {
+	return eval.StartEvaluator(g.ledger, hdr,
+		eval.EvaluatorOptions{
+			PaysetHint:          paysetHint,
+			Generate:            true,
+			Validate:            false,
+			MaxTxnBytesPerBlock: 0,
+			Tracer:              nil,
+		})
+}
+
+func (g *generator) evaluateBlock(hdr bookkeeping.BlockHeader, txGroups [][]txn.SignedTxnWithAD, paysetHint int) (*ledgercore.ValidatedBlock, uint64 /* txnCount */, error) {
+	eval, err := g.startEvaluator(hdr, paysetHint)
+	if err != nil {
+		return nil, 0, fmt.Errorf("could not start evaluator: %w", err)
+	}
+	for i, txGroup := range txGroups {
+		err := eval.TransactionGroup(txGroup)
+		if err != nil {
+			return nil, 0, fmt.Errorf("could not evaluate transaction group %d: %w", i, err)
+		}
+	}
+	lvb, err := eval.GenerateBlock()
+	return lvb, eval.TestingTxnCounter(), err
+}
```

Comment:
Nice, this doesn't seem so bad.

Reply:
Yeah... I don't know why I was so resistant to this approach. I was trying to avoid modifying how we used the ledger at all costs. But I saw the light thanks to @algorandskiy .